http-clients/src/main/java/org/triplea/http/client/web/socket/WebSocketConnection.java

```diff
@@ -95,53 +109,71 @@ void close() {
   }
 
   /**
-   * Initiates a websocket connection. Must be called before {@link #sendMessage(String)}. This
-   * method blocks until the connection has either successfully been established, or the connection
-   * failed.
+   * Initiates a non-blocking websocket connection.
    *
-   * @throws CouldNotConnect If the connection fails
+   * @param errorHandler Invoked if there is a failure to connect.
+   * @throws IllegalStateException Thrown if connection is already open (eg: connect called twice).
+   * @throws IllegalStateException Thrown if connection has been closed (ie: 'close()' was called)
    */
-  void connect() {
+  // Suppression: Ignoring future return values ignores any exceptions thrown by the code
+  // that completes the future. In this case we ignore the returned future from 'whenComplete'
+  // to avoid blocking. We would not expect the ping sender start or error handler to throw
+  // exceptions.
+  @SuppressWarnings("FutureReturnValueIgnored")
+  void connect(final Consumer<String> errorHandler) {
     Preconditions.checkState(!client.isOpen());
     Preconditions.checkState(!closed);
-    try {
-      client.connect();
-      pingSender.start();
-    } catch (final Exception e) {
-      pingSender.cancel();
-      throw new CouldNotConnect(serverUri);
-    }
+
+    connectAsync()
+        .whenComplete(
+            (connected, throwable) -> {
+              if (connected && throwable == null) {
+                pingSender.start();
+              } else {
+                errorHandler.accept("Failed to connect to: " + serverUri);
+              }
+            });
```

Comment:
Well technically we could get an error here, maybe the errorHandler throws some exception, or the ping sender has some issue.
In both cases it would just silently fail.

> We would not expect the ping sender start or error handler to throw exceptions.

I don't think exceptions are expected at any time really unless you're going for a pythonic EAFP style:
https://docs.python.org/3/glossary.html#term-eafp 🤷‍♂ 

This is an perfectly fine example to use `CompletableFutureUtils`

Reply:
`CompletableFutureUtils` would potentially just add another `whenComplete` clause. The error handling would be the same, so we'd except a failure there just the same and would not really be adding much.

For refernece:
```
    future.whenComplete(
        (result, ex) -> {
          if (ex != null) {
            exceptionHandler.accept(ex);
          }
        });
```

It's also interesting that we have a suppress warning clause, perhaps the `whenComplete` can be transformed into an `exceptionally` clause in `CompletableFutureUtils` (in which case, maybe the helper class would not be needed at all).

e48b6d909 updates this and removes this second warning suppression.